File: helix_desktop/Helix_NetduinoGo/Helix_Netduino/GA.cs
using System;

namespace Helix
{
    class GA
    {
        private const int __population_size__ = 20;
        public const int __loci_count__ = 30;
        private const int __max_colorfulness__ = __loci_count__ * 255 * 2;

        private static double per_site_mutation_rate = 0.01;
        private int generation;

        protected static Random random_generator = new Random();

        public Organism[] Population = new Organism[__population_size__];
        private Organism[] Candidates = new Organism[__population_size__];

        public GA()
        {
            random_generator = new Random();
            generation = 0;
            init();
        }

        public void init()
        {
            init_random_population();
        }

        public static bool is_mut()
        {
            if (per_site_mutation_rate * 1000 < random_generator.Next(999))
                return true;
            return false;
        }

        public void evolve()
        {
            select_candidates();
            mutate_candidates();
            recombine_candidates_into_population();
            sort_population();

            generation++;
        }

        private void sort_population()
        {
            comb_sort(Population, __population_size__);
        }

        void comb_sort(Organism[] input, int n) 
        {  
            Organism swap;
            int i, gap = n;
            bool swapped = false;
 
            while ((gap > 1) || swapped) {
                if (gap > 1) {
                    gap = (int)((double)gap / 1.247330950103979);
                }
 
                swapped = false;
 
                for (i = 0; gap + i < n; ++i) {                
                    if (input[i].inverse_fitness - input[i + gap].inverse_fitness > 0) {
                        swap = input[i];
                        input[i] = input[i + gap];
                        input[i + gap] = swap;
                    
                        //swap_genomes(i, i + gap); // swap
                
                        swapped = true;
                    }
                }
            }
        }


        private void recombine_candidates_into_population()
        {
            for (int i = 0; i < __population_size__; i += 2)
            {
                Population[i] = new Organism( Candidates[ random_generator.Next(__population_size__ - 1) ] );
                Population[i+1] = new Organism( Candidates[ random_generator.Next(__population_size__ - 1) ] );

                Organism.recombine(Population[i], Population[i + 1]);
            }
        }

        private void mutate_candidates()
        {
            foreach (Organism aOrg in Candidates)
                aOrg.mutate();
        }

        private void select_candidates()
        {
//            select_candidates__elitism();
            select_candidates__hybrid();
        }

        // select all candidates from the top 50%
        private void select_candidates__elitism()
        {
            for (int i = 0; i < __population_size__; i++)
            {
                Candidates[i] = new Organism(
                    Population[random_generator.Next((int)(__population_size__ * .5))]);
            }
        }

        private void select_candidates__hybrid()
        {
            // 50% are drawn from top 20%
            int target = 0;
            for (int i = 0; i < (int)(__population_size__ * .5); i++, target++)
            {
                int popCandidate = random_generator.Next((int)(__population_size__ * .2) - 1);
                Candidates[target] = new Organism(
                    Population[popCandidate]);
            }
            // 30% are drawn from middle 60%
            for (int i = 0; i < (int)(__population_size__ * .3); i++, target++)
            {
                Candidates[target] = new Organism(
                    Population[random_generator.Next((int)(__population_size__ * .6)) + 
                    (int)(__population_size__ * .2) - 1]);
            }
            // 20% are drawn from bottom 20%
            for (int i = 0; i < (int)(__population_size__ * .2); i++, target++)
            {
                Candidates[target] = new Organism(
                    Population[random_generator.Next((int)(__population_size__ * .2)) +
                    (int)(__population_size__ * .8) - 1]);
            }
        }

        public void reset_ga()
        {
            per_site_mutation_rate = 0.01;
            init_random_population();
        }

        private void init_random_population()
        {
            for (int i = 0; i < __population_size__; i++)
            {
                Population[i] = new Organism();
                Population[i].randomize();
            }

        }

        public struct Trait
        {
            public byte r, g, b;

            public Trait(byte rVal, byte gVal, byte bVal)
            {
                r = rVal;
                g = gVal;
                b = bVal;
            }

            public void randomize()
            {
                r = get_random();
                g = get_random();
                b = get_random();
            }

            private byte get_random()
            {
                return (byte)(GA.random_generator.Next(255));
            }

            public int calculate_colorfulness()
            {
                return System.Math.Abs(r - g) + System.Math.Abs(g - b) + System.Math.Abs(b - r);
            }

            public static int calculate_difference(Trait aTraitA, Trait aTraitB)
            {
                return System.Math.Abs(aTraitA.r - aTraitB.r) +
                    System.Math.Abs(aTraitA.g - aTraitB.g) +
                    System.Math.Abs(aTraitA.b - aTraitB.b);
            }

            public void mutate()
            {
                if (is_mut())
                    r = get_random();
                if (is_mut())
                    b = get_random();
                if (is_mut())
                    b = get_random();
            }
        }

        public class Organism : IComparable
        {
            public Trait[] Traits = new Trait[__loci_count__];
            public int inverse_fitness;

            public Organism()
            {
                inverse_fitness = int.MaxValue;
            }

            public Organism( Organism aOriginal ) 
            {
                for (int i = 0; i < __loci_count__; i++)
                    Traits[i] = aOriginal.Traits[i];
                //Array.Copy(Traits, aOriginal.Traits, __loci_count__);
                inverse_fitness = aOriginal.inverse_fitness;
            }

            public void randomize()
            {
                for (int i = 0; i < __loci_count__; i++)
                {
                    Traits[i].randomize();
                }
                calculate_fitness();
            }

            public void calculate_fitness()
            {
                int sumDiff = 0;
                for (int i = 0; i < __loci_count__ - 1; i++)
                {
                    sumDiff += Trait.calculate_difference(Traits[i], Traits[i+1]);
                }
                sumDiff += Trait.calculate_difference(Traits[0], Traits[__loci_count__ - 1]);

                int sumColor = 0;
                foreach (Trait aTrait in Traits)
                {
                    sumColor += aTrait.calculate_colorfulness();
                }
                int invSumColor = __max_colorfulness__ - sumColor;
                inverse_fitness = sumDiff + invSumColor;
            }

            int IComparable.CompareTo(object obj)
            {
                if (obj is Organism) {
                    Organism aOrg = (Organism) obj;
                    if (inverse_fitness < aOrg.inverse_fitness)
                        return -1;
                    if (inverse_fitness > aOrg.inverse_fitness)
                        return 1;
                    return 0;
                }
                throw new ArgumentException("object is not an Organism");    
            }

            internal void mutate()
            {
                foreach (Trait aTrait in Traits)
                {
                    aTrait.mutate();
                }
            }
            internal void mutate(int index)
            {
                Traits[index].mutate();
            }

            internal static void recombine(Organism organism, Organism organism_2)
            {
                int posA = random_generator.Next(__loci_count__ - 1);
                int posB = random_generator.Next(__loci_count__ - 1);

                bool recomb = false;
                for (int i = 0; i < __loci_count__; i++)
                {
                    if (i == posA || i == posB)
                        recomb = !recomb;

                    if (recomb) // swap
                    {
                        Trait lTmp = organism.Traits[i];
                        organism.Traits[i] = organism_2.Traits[i];
                        organism_2.Traits[i] = lTmp;
                    }
                }

                organism.calculate_fitness();
                organism_2.calculate_fitness();
            }
        }
    }
}

File: axis/axis/GA.cpp
//
//  GA.cpp
//  Helix
//
//  Created by Rosangela Canino-Koning on 1/28/13.
//  Modified by Rosangela Canino-Koning on 5/9/24.
//  Copyright (c) 2024 Batch6. All rights reserved.
//

#include "GA.h"

GA::GA()
{
    
    // init the content of the Populations Array
    int PopulationSizes[] = __POPULATION_SIZES__;
    
    int startIndex = 0;
    for (int i = 0; i < __POPULATIONS__; i++)
    {
        Populations[i].StartIndex = startIndex;
        Populations[i].Count = PopulationSizes[i];
        
        startIndex += PopulationSizes[i];
    }
    
    //current_mutation_rate = per_site_mutation_rate;
    
    use_heredity = true;
    use_selection = true;
    generation = true; 
}

double GA::per_site_mutation_rate = __DEFAULT_MUTATION_RATE__;
double GA::red_bias = 1;
double GA::green_bias = 1;
double GA::blue_bias = 1;


void GA::increase_mutation_rate()
{
  per_site_mutation_rate = per_site_mutation_rate >= 1 ? 1 : per_site_mutation_rate + 0.01;
  DEBUG_INFO("Mutation Rate: %f", per_site_mutation_rate);
}

void GA::decrease_mutation_rate()
{
  per_site_mutation_rate = per_site_mutation_rate <= 0 ? 0 : per_site_mutation_rate - 0.01;
  DEBUG_INFO("Mutation Rate: %f", per_site_mutation_rate);

}

void GA::increase_bias(bool r, bool g, bool b)
{
  if (r)
    red_bias *= 1.1;
  if (g)
    green_bias *= 1.1;
  if (b)
    blue_bias *= 1.1;

    DEBUG_INFO("Bias: r %f, g %f, b %f", red_bias, green_bias, blue_bias);
}
void GA::init()
{
    // init the random seed
    randomSeed(analogRead(0));
    
    init_random_population();
}

void GA::reset_ga()
{
    red_bias = 1;
    green_bias = 1;
    blue_bias = 1;

    per_site_mutation_rate = __DEFAULT_MUTATION_RATE__;

    init();
}

void GA::evolve()
{
    //migrate(); // if there is any migration between populations, do it now.
    
    for (int i = 0; i < __POPULATIONS__; i++)
    {
        select_candidates(i);
        mutate_candidates(i);
        reproduce_candidates(i);
        sort_population(i);
    }
    generation++;
}

void GA::sort_population(int popIndex)
{
    comb_sort_pop( Populations[popIndex].StartIndex, Populations[popIndex].Count );
}

void GA::comb_sort_pop(int start, int n)
{
    Organism swap;
    int i, gap = n;
    bool swapped = false;
    
    while ((gap > 1) || swapped) {
        if (gap > 1) {
            gap = (int)((double)gap / 1.247330950103979);
        }
        
        swapped = false;
        
        for (i = 0; gap + i < n; ++i) {
            if (Organisms[i + start].inverse_fitness - Organisms[i + start + gap].inverse_fitness > 0) {
                swap.clone( Organisms[i + start] );
                Organisms[i + start].clone( Organisms[i + start + gap] );
                Organisms[i + start + gap].clone( swap );
                
                swapped = true;
            }
        }
    }
}

void GA::debug_all()
{

    for (int i = 0; i < __TOTAL_ORGANISMS__; i++)
    {
        Organisms[i].debug();
    }
    DEBUG_VERBOSE("\n");
}

void GA::debug_candidates()
{

    for (int i = 0; i < __LARGEST_POPULATION__; i++)
    {
        Candidates[i].debug();
    }
    DEBUG_VERBOSE("\n");
    
}

void GA::debug_population(int popIndex)
{
    DEBUG_VERBOSE("Pop %d  - Best Fitness: %f", popIndex, Organisms[Populations[popIndex].StartIndex].inverse_fitness);    
    for (int i = 0; i < Populations[popIndex].Count; i++)
    {
        Organisms[Populations[popIndex].StartIndex + i].debug();
    }    
    
}

void GA::init_random_population()
{
    for (int i = 0; i < __TOTAL_ORGANISMS__; i++)
    {
        Organisms[i].randomize();
    }
}

void GA::randomize_population(int popIndex)
{
    for (int i = Populations[popIndex].StartIndex; i < Populations[popIndex].StartIndex + Populations[popIndex].Count; i++)
    {
        Organisms[i].randomize();
    }
}

bool GA::is_mut()
{
    int randval = random(0,999);
    if (per_site_mutation_rate * 1000 > randval)
    {        
        return true;
    }
    return false;
}

void GA::select_candidates(int popIndex)
{
    if (use_selection)
        //select_candidates__hybrid(popIndex);
        select_candidates__hybrid__evoluminator(popIndex);
    else
        select_candidates__random(popIndex);
}

void GA::mutate_candidates(int popIndex)
{
    for (int org = 0; org < Populations[popIndex].Count; org++)
    {
        Candidates[org].mutate();
    }
}

void GA::reproduce_candidates(int popIndex)
{
    if (use_heredity)
        recombine_candidates_into_population(popIndex);
    else
        randomize_population(popIndex);
}

void GA::recombine_candidates_into_population(int popIndex)
{    
    for (int i = 0; i < Populations[popIndex].Count; i++)
    {
        int candidate1 = random(0, Populations[popIndex].Count);
        int candidate2 = random(0, Populations[popIndex].Count);
        
        int childIndex = Populations[popIndex].StartIndex + i;
        
        BirthChamber[0].clone( Candidates[ candidate1 ] );
        BirthChamber[1].clone( Candidates[ candidate2 ] );
        
        Organism::recombine(BirthChamber[0], BirthChamber[1]);
        
        int randomindex = random(0, 2);
        //Debug::debugln(randomindex);
        Organisms[childIndex].clone( BirthChamber[ randomindex ] );

    }
}

// select randomly from the population
void GA::select_candidates__random(int popIndex)
{
    for (int i = 0; i < Populations[popIndex].Count; i++)
    {
        int candidateIndex = random(0, Populations[popIndex].Count);
        Candidates[i].clone(Organisms[Populations[popIndex].StartIndex + candidateIndex]);
    }
}

// select all candidates from the top 50%
void GA::select_candidates__elitism(int popIndex)
{
    for (int i = 0; i < Populations[popIndex].Count; i++)
    {
        int candidateIndex = random(0, Populations[popIndex].Count * .5); // only the top 50%
        Candidates[i].clone(Organisms[Populations[popIndex].StartIndex + candidateIndex]);
    }
}

void GA::select_candidates__hybrid__evoluminator(int popIndex)
{
    int popsize = Populations[popIndex].Count;
    int offset = Populations[popIndex].StartIndex;
    
    int l = 0, r = 0, random_index = 0, target = 0;
    
    // 50% of candidates are drawn from the top 25%
    l = 0;
    r = popsize * .25;
    for (int i = 0; i < (popsize * .5); i++) {
        random_index = random(l, r + 1) + offset;
        Candidates[target++].clone(Organisms[random_index]);
    }    
    
    // 30% of candidates are drawn from the middle 50%
    l = popsize * .25;
    r = popsize * .75;
    for (int i = 0; i < (popsize * .3); i++) {
        random_index = random(l, r + 1) + offset;
        Candidates[target++].clone(Organisms[random_index]);
    }
    
    // 20% of candidates are drawn from the bottom 25%
    l = popsize * .75;
    r = popsize - 1;
    for (int i = 0; i < (popsize * .2); i++) {
        if (target >= popsize) // avoid an overrun
            break;
        
        random_index = random(l, r + 1) + offset;
        Candidates[target++].clone(Organisms[random_index]);
    }
}

void GA::select_candidates__hybrid(int popIndex)
{
    // 50% are drawn from top 20%
    int target = 0;
    for (int i = 0; i < (int)(Populations[popIndex].Count * .5); i++, target++)
    {
        int popCandidate = random(0, Populations[popIndex].Count * .2) - 1;
        Candidates[target].clone(Organisms[Populations[popIndex].StartIndex + popCandidate]);
    }
    // 30% are drawn from middle 60%
    for (int i = 0; i < (int)(Populations[popIndex].Count * .3); i++, target++)
    {
        int popCandidate = random(0, (int)(Populations[popIndex].Count * .6)) +
            (int)(Populations[popIndex].Count * .2) - 1;
        Candidates[target].clone(Organisms[Populations[popIndex].StartIndex + popCandidate]);
    }
    // 20% are drawn from bottom 20%
    for (int i = 0; i < (int)(Populations[popIndex].Count * .2); i++, target++)
    {
        int popCandidate = random(0, (int)(Populations[popIndex].Count * .2)) +
            (int)(Populations[popIndex].Count * .8) - 1;
        Candidates[target].clone(Organisms[Populations[popIndex].StartIndex + popCandidate]);
    }
}

////////////////////////
// ORGANISMS SUB CLASS
////////////////////////
GA::Organism::Organism()
{
    randomize();
    calculate_fitness();
}

void GA::Organism::calculate_fitness()
{
    // sum of differences across the genome
    int sumDiff = 0;
    for (int i = 0; i < __LOCUS_COUNT__ - 1; i++)
    {
        sumDiff += Trait::calculate_difference(Traits[i], Traits[i+1]);
    }
    sumDiff += Trait::calculate_difference(Traits[0], Traits[__LOCUS_COUNT__ - 1]);
    
    // sum of the colorfulness across the genome
    int sumColor = 0;
    for (int i = 0; i < __LOCUS_COUNT__ - 1; i++)
    {
        sumColor += Traits[i].calculate_colorfulness();
    }
    int invSumColor = _MAX_COLORFULNESS_ - sumColor;
    
    inverse_fitness = (sumDiff * 4) + invSumColor;
}

void GA::Organism::randomize()
{
    for (int i = 0; i < __LOCUS_COUNT__; i++)
    {
        Traits[i].randomize();
    }
    calculate_fitness();
}

int GA::Organism::CompareTo(Organism & aOrg)
{
    if (inverse_fitness < aOrg.inverse_fitness)
        return -1;
    if (inverse_fitness > aOrg.inverse_fitness)
        return 1;
    return 0;
}

void GA::Organism::mutate()
{
    for (int i = 0; i < __LOCUS_COUNT__; i++)
    {
        Traits[i].mutate();
    }
}

void GA::Organism::mutate(int index)
{
    Traits[index].mutate();
}

void GA::Organism::recombine(Organism & organism, Organism & organism_2)
{
    int posA; //= random_generator.Next(__loci_count__ - 1);
    int posB; //= random_generator.Next(__loci_count__ - 1);
    
    bool recomb = false;
    for (int i = 0; i < __LOCUS_COUNT__; i++)
    {
        if (i == posA || i == posB)
            recomb = !recomb;
        
        if (recomb) // swap
        {
            Trait lTmp = organism.Traits[i];
            organism.Traits[i] = organism_2.Traits[i];
            organism_2.Traits[i] = lTmp;
        }
    }
    
    organism.calculate_fitness();
    organism_2.calculate_fitness();
}

void GA::Organism::clone(Organism & aOriginal)
{
    for (int i = 0; i < __LOCUS_COUNT__; i++)
    {
        Traits[i].clone(aOriginal.Traits[i]);
    }
    inverse_fitness = aOriginal.inverse_fitness;
}

void GA::Organism::debug()
{    
    DEBUG_VERBOSE("%f", inverse_fitness);
    DEBUG_VERBOSE(":");
   
    for (int i = 0; i < __LOCUS_COUNT__; i++)
    {
        DEBUG_VERBOSE(" ");
        Traits[i].debug();
    }
    
    DEBUG_VERBOSE("\n"); 
}
